File: server/src/GridSync.Core/Rows/FractionalIndex.cs
namespace GridSync.Core.Rows;

/// <summary>
/// Keys that sort between any two neighbours, so a row can be inserted anywhere without renumbering
/// the rows around it. This is what lets two people insert rows at the same time: each picks a key
/// between the same two neighbours, and no other row has to move.
///
/// A key is a string of base-62 digits ("0-9A-Za-z", which is also their ASCII order, so plain
/// ordinal string comparison sorts keys correctly), read as a fraction: "V" is about one half,
/// "V5" a little more. Two rules keep it sound: a key never ends in "0" (it would equal the key
/// without it, "V0" == "V", so two spellings of one position), and Between needs before &lt; after.
///
/// Mirrors client/src/app/rows/fractional-index.ts. Both must give identical keys, so the shared
/// vectors in spec/fractional-index-vectors.json are run by both test suites.
/// </summary>
public static class FractionalIndex
{
    public const string Digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

    private const int Base = 62;

    /// <summary>Width of a base row's key. Four digits reach 62^4 = 14.7 million rows.</summary>
    private const int BaseRowWidth = 4;

    /// <summary>
    /// A key strictly between two others. Null <paramref name="before"/> (or empty) means "before
    /// everything", null <paramref name="after"/> means "after everything".
    /// </summary>
    public static string Between(string? before, string? after)
    {
        before ??= string.Empty;
        Validate(before, nameof(before), allowEmpty: true);
        if (after is not null) Validate(after, nameof(after), allowEmpty: false);

        if (after is not null && string.CompareOrdinal(before, after) >= 0)
            throw new ArgumentException($"'{before}' must sort before '{after}'.");

        return Midpoint(before, after);
    }

    /// <summary>
    /// The key of the Nth row a sheet starts with. The sheet's initial rows are implicit (100,000
    /// records nobody has touched are not worth storing), so their keys are computed, not stored:
    /// the index in base 62, padded to a fixed width so they sort in index order, plus a "V" so the
    /// key never ends in "0" and there is always room to insert between two neighbouring rows.
    /// </summary>
    public static string ForBaseRow(int index)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index);

        var digits = new char[BaseRowWidth];
        var rest = index;
        for (var i = BaseRowWidth - 1; i >= 0; i--)
        {
            digits[i] = Digits[rest % Base];
            rest /= Base;
        }

        if (rest != 0) throw new ArgumentOutOfRangeException(nameof(index), "Too many rows for a four digit key.");
        return new string(digits) + "V";
    }

    /// <summary>True when <paramref name="key"/> is a well formed key: non-empty, only key digits, not ending in "0".</summary>
    public static bool IsValid(string? key)
    {
        if (string.IsNullOrEmpty(key) || key[^1] == '0') return false;
        foreach (var c in key)
        {
            if (Digits.IndexOf(c) < 0) return false;
        }

        return true;
    }

    private static string Midpoint(string a, string? b)
    {
        if (b is not null)
        {
            // Keys that share a prefix share it in the answer: only the part after it matters. A
            // key shorter than the prefix counts as padded with zeros ("1" is "100...").
            var n = 0;
            while (n < b.Length && (n < a.Length ? a[n] : '0') == b[n]) n++;
            if (n > 0) return b[..n] + Midpoint(n < a.Length ? a[n..] : string.Empty, b[n..]);
        }

        var digitA = a.Length > 0 ? Digits.IndexOf(a[0]) : 0;
        var digitB = b is not null ? Digits.IndexOf(b[0]) : Base;

        if (digitB - digitA > 1) return Digits[(digitA + digitB + 1) / 2].ToString(); // room between the first digits

        // The first digits are neighbours (or equal, only possible when before is empty), so a one
        // digit key cannot fit between. Either shorten to b's first digit, which is above a and a
        // proper prefix of b (so below it), or keep a's digit and go between the rest of a and "the end".
        if (b is not null && b.Length > 1) return b[..1];
        return Digits[digitA] + Midpoint(a.Length > 0 ? a[1..] : string.Empty, null);
    }

    private static void Validate(string key, string name, bool allowEmpty)
    {
        if (key.Length == 0)
        {
            if (allowEmpty) return;
            throw new ArgumentException("A key cannot be empty.", name);
        }

        foreach (var c in key)
        {
            if (Digits.IndexOf(c) < 0) throw new ArgumentException($"'{key}' has a character outside 0-9A-Za-z.", name);
        }

        if (key[^1] == '0') throw new ArgumentException($"'{key}' ends in 0, which is another spelling of a shorter key.", name);
    }
}

File: client/src/app/rows/fractional-index.ts
/**
 * Keys that sort between any two neighbours, so a row can be inserted anywhere without renumbering
 * the rows around it. Mirrors server/src/GridSync.Core/Rows/FractionalIndex.cs: read the long
 * comment there for the idea. In short, a key is a string of base-62 digits ("0-9A-Za-z", which is
 * also their ASCII order) read as a fraction, it never ends in "0", and plain string comparison
 * orders keys. Both implementations must give identical keys; spec/fractional-index-vectors.json is
 * run by both test suites.
 */
export const DIGITS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';

const BASE = 62;
const BASE_ROW_WIDTH = 4;

/** Ordinal comparison: UTF-16 code unit order, which for these ASCII keys is the order that matters. */
export function compareKeys(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/** A key strictly between two others. Null or empty `before` means "before everything"; null `after` means "after everything". */
export function between(before: string | null, after: string | null): string {
  const start = before ?? '';
  validate(start, 'before', true);
  if (after !== null) validate(after, 'after', false);

  if (after !== null && compareKeys(start, after) >= 0) throw new Error(`'${start}' must sort before '${after}'.`);

  return midpoint(start, after);
}

/**
 * The key of the Nth row a sheet starts with. Initial rows are implicit, so their keys are
 * computed, not stored: the index in base 62, padded to a fixed width so they sort in index
 * order, plus a "V" so the key never ends in "0" and there is room between neighbouring rows.
 */
export function forBaseRow(index: number): string {
  if (!Number.isInteger(index) || index < 0) throw new Error('A row index must be a non-negative integer.');

  const digits: string[] = [];
  let rest = index;
  for (let i = 0; i < BASE_ROW_WIDTH; i++) {
    digits.unshift(DIGITS.charAt(rest % BASE));
    rest = Math.floor(rest / BASE);
  }

  if (rest !== 0) throw new Error('Too many rows for a four digit key.');
  return digits.join('') + 'V';
}

function midpoint(a: string, b: string | null): string {
  if (b !== null) {
    // Keys that share a prefix share it in the answer. A key shorter than the prefix counts as padded with zeros.
    let n = 0;
    while (n < b.length && (n < a.length ? a.charAt(n) : '0') === b.charAt(n)) n++;
    if (n > 0) return b.substring(0, n) + midpoint(n < a.length ? a.substring(n) : '', b.substring(n));
  }

  const digitA = a.length > 0 ? DIGITS.indexOf(a.charAt(0)) : 0;
  const digitB = b !== null ? DIGITS.indexOf(b.charAt(0)) : BASE;

  if (digitB - digitA > 1) return DIGITS.charAt(Math.floor((digitA + digitB + 1) / 2));

  if (b !== null && b.length > 1) return b.substring(0, 1);
  return DIGITS.charAt(digitA) + midpoint(a.length > 0 ? a.substring(1) : '', null);
}

function validate(key: string, name: string, allowEmpty: boolean): void {
  if (key.length === 0) {
    if (allowEmpty) return;
    throw new Error(`${name} cannot be empty.`);
  }

  for (const c of key) {
    if (DIGITS.indexOf(c) < 0) throw new Error(`'${key}' has a character outside 0-9A-Za-z.`);
  }

  if (key.charAt(key.length - 1) === '0') throw new Error(`'${key}' ends in 0, which is another spelling of a shorter key.`);
}
